registration/approval.py

```diff
@@ -51,33 +48,61 @@ def _send_mail(application, template_name, subject):
         from_email=settings.BETATEST_EMAIL_SENDER,
         recipient_list=(application.user.email,),
     )
+    # Using a separate call to hide internal email
+    send_mail(
+        subject=subject,
+        message=message,
+        from_email=settings.BETATEST_EMAIL_SENDER,
+        recipient_list=(settings.BETATEST_EMAIL_INTERNAL,),
+    )
 
 
-def accept_application(application):
+def accept_application(application, appserver):
     """Accept a beta test application.
 
     This helper function verifies that an AppServer for this application has been successfully
     launched, activates it and sends an email to the user to notify them that their instance is
     ready.
     """
-    assert application.instance is not None, 'No instance provisioned yet.'
-    appserver = application.instance.active_appserver
-    assert appserver is not None, 'The instance does not have an active AppServer yet.'
-    assert appserver.status == AppServer.Status.Running, 'The AppServer is not running yet.'
-    _send_mail(application, 'registration/welcome_email.txt', BETATEST_WELCOME_SUBJECT)
+    if application.instance is None:
+        raise ApplicationNotReady('No instance provisioned yet.')
+
+    if appserver is None:
+        raise ApplicationNotReady('The instance does not have an active AppServer yet.')
+    if appserver.status != AppServer.Status.Running:
+        raise ApplicationNotReady('The AppServer is not running yet.')
+
+    _send_mail(application, 'registration/welcome_email.txt', settings.BETATEST_WELCOME_SUBJECT)
     application.status = BetaTestApplication.ACCEPTED
     application.save()
 
 
-def reject_application(application):
-    """Reject a beta test application.
+@receiver(appserver_spawned)
+def on_appserver_spawned(sender, **kwargs):
+    """
+    Monitor spawning of new appservers, to send the welcome email once it is ready.
+    """
+    instance = kwargs['instance']
+    appserver = kwargs['appserver']
+    application = instance.betatestapplication_set.first()  # There should only be one
+
+    if not application:
+        return
+
+    elif application.status != BetaTestApplication.PENDING:
+        return
```

Comment:
This and the above can be combined into a single statement.

Reply:
👍